torchao/prototype/mx_formats/custom_cast.py

```diff
@@ -1080,3 +1082,498 @@ def _(uint8_data):
+        )
+        # add offset for inner blocks
+        factor = ROW_TILE_SIZE // INNER_BLOCK_SIZE
+
+        # needs better name
+        jump_vals_per_col = (n_rows - ROW_TILE_SIZE) // INNER_BLOCK_SIZE
+        # tl.device_print("jump_vals_per_col", jump_vals_per_col)
+
+        # [0, 1, 2, 3, 4, 5, 6, 7] -> [0, 1, 4, 5, 8, 9, 12, 13]
+        col_scale_indices = col_scale_indices + (
+            tl.floor(col_scale_indices / factor) * jump_vals_per_col
+        ).to(tl.int32)
+        # tl.static_print(col_scale_indices)
+        # tl.device_print("indices", col_scale_indices)
+
+        # TODO(future): mask
+        tl.store(col_scale_start_ptr + col_scale_indices, col_scale_e8m0)
+
+    def to_mxfp8_dim1(x, inner_block_size=32):
+        """
+        Input:
+        * `x` - input tensor, in row major memory layout
+        * `inner_block_size` - size of tiles to scale across, default is 32 for MX recipes
+
+        Output:
+        * `output_col_major`: the `float8_e4m3fn` values of `x` cast to mxfp8 across dim1
+        * `col_scale`: the `e8m0` values of `x_scale` used to cast `x` to mxfp8 across dim1
+        """
+        assert x.is_contiguous(), "`x` must be contiguous"
+        assert x.dtype == torch.bfloat16
+        # Get tensor shape
+        n_rows, n_cols = x.shape
+
+        # can be autotuned
+        # row_tile_size = 32
+        row_tile_size = 128
+        # row_tile_size = 4
+
+        assert row_tile_size >= inner_block_size
+
+        # TODO autotune col_tile_size
+        # input 16k by 16k
+        # triton scaling mostly commented out
+        # row_tile_size=256, col_tile_size=64: 3.47 TB/s
+        # TODO(next): make calculations work with ^
+        col_tile_size = 128
+        # col_tile_size = 4
+
+        # Create output tensors
+        output_col_major = torch.empty(
+            (n_cols, n_rows), dtype=torch.float8_e4m3fn, device=x.device
+        )
+
+        # Create scale tensors
+        # TODO(before land): switch back to empty
+        col_scale = torch.empty(
+            n_cols, n_rows // inner_block_size, dtype=torch.uint8, device=x.device
+        )
+
+        # Calculate grid dimensions based on tile size
+        grid_rows = triton.cdiv(n_rows, row_tile_size)
+        grid_cols = triton.cdiv(n_cols, col_tile_size)
+
+        # inner_block_size = 32
+        rename_me_tile_size = row_tile_size * col_tile_size // inner_block_size
+
+        # Launch the kernel
+        to_mxfp8_dim1_kernel[(grid_rows, grid_cols)](
+            x_ptr=x,
+            output_col_major_ptr=output_col_major,
+            col_scale_ptr=col_scale,
+            n_rows=n_rows,
+            n_cols=n_cols,
+            COL_TILE_SIZE=col_tile_size,
+            ROW_TILE_SIZE=row_tile_size,
+            RENAME_ME_TILE_SIZE=rename_me_tile_size,
+            INNER_BLOCK_SIZE=inner_block_size,
+        )
+
+        return (
+            output_col_major.t(),
+            col_scale.reshape(-1, 1).view(torch.float8_e8m0fnu),
```

Comment:
Because the triton is just accessing `data_ptr` you could initialize these as the correct layout if you wanted.